src/osp/Active/SysPhysics.cpp

```diff
@@ -1,2 +1,72 @@
 #include "SysPhysics.h"
-#include "basic.h"
+#include "SysSceneGraph.h"
+
+using namespace osp;
+using namespace osp::active;
+
+void SysPhysics::calculate_subtree_mass_center(
+        acomp_storage_t<ACompTransform> const&  rTf,
+        ACtxPhysics&                            rCtxPhys,
+        ACtxSceneGraph&                         rScnGraph,
+        ActiveEnt                               root,
+        Vector3&                                rMassPos,
+        float&                                  rTotalMass,
+        Matrix4 const&                          currentTf)
+{
+    for (ActiveEnt const child : SysSceneGraph::children(rScnGraph, root))
```

Comment:
Cool thing about the hierarchy here is that subtrees are contiguous int arrays. This recursive function is actually a linear array iteration.